extensions/ql-vscode/src/remote-queries/analyses-results-manager.ts

```diff
@@ -106,7 +106,8 @@ export class AnalysesResultsManager {
     const resultsForQuery = this.internalGetAnalysesResults(queryId);
     resultsForQuery.push(analysisResults);
     this.analysesResults.set(queryId, resultsForQuery);
-    void publishResults(resultsForQuery);
+    void publishResults([...resultsForQuery]);
+    const pos = resultsForQuery.length - 1;
```

Comment:
I've always like immutable.js for things like this, but I don't like introducing it into an ongoing project. Maybe it's not such a bad idea in this case.

We do need to be careful. Now that we are saving data across restarts, any change in format will cause error when there is an upgrade. It's not a _terrible_ situation since the worst case is that we need to delete some old queries on upgrade.

I should add version number to each saved artifact and the extension can read this and upgrade it appropriately when we change versions.

Reply:
I'd be up for using immutable.js or any other way to enforce immutability.

Agree about versions - we'll need to be careful. Perhaps the overall query history store has a version? We can always deal with it the first time we need a new version I think.